### PR TITLE
minimal patch to fix Windows compilation issues

### PR DESCRIPTION
**Pull request by wkpark**
manually cherry-picked from PR #788 and PR #229 and cleaned up by me

Original work done by @Jamezo97 and @acpopescu

### CODE REVIEW

**Comment by wkpark** [comment]
+ cmake fix + github workflows with cuda 12.1, 11.8 available at. (github login and scroll down to download artifacts)
https://github.com/wkpark/bitsandbytes/actions/runs/7030297936

**Comment by wkpark** [comment]
cmake + github workflows + cuda 12.1, 11.8 + python 3.10, 3.11 + windows, ubuntu build matrix available.
https://github.com/wkpark/bitsandbytes/actions/runs/7165411186/job/19507307550

**Comment by TimDettmers** [comment]
This is awesome, thanks for your contribution! We will discuss this internally and get back to you. @Titus-von-Koeller @younesbelkada please remind me that we talk about this.

File: csrc/cpu_ops.cpp [hunk 1]
```diff
@@ -23,15 +27,19 @@ void quantize_cpu(float *code, float *A, float *absmax, unsigned char *out, long
...
+    BinAlgo<INSTR_SET, float, Direct2> bin_searcher(code, elements_code);
 
     int thread_wave_size = 256;
     // we chunk the thresds into waves of 256 since the max limit is
     // between 16k and 64k on Linux (we reach this when running BLOOM-176B with a large batch size)
     for(long long offset = 0; offset < num_blocks; offset+=thread_wave_size)
     {
       long long valid_chunks = num_blocks - offset >= thread_wave_size ? thread_wave_size : num_blocks - offset;
+#ifdef _WIN32
+      std::thread *threads = (std::thread *) malloc(sizeof(std::thread) * valid_chunks);
```
Comment (rickardp) [inline csrc/cpu_ops.cpp:39]: Any reason not to use std::thread everywhere?

Comment (wkpark) [reply csrc/cpu_ops.cpp:39]: > Any reason not to use std::thread everywhere?

I just leaving this as the original patch creator intended.

**Comment by Titus-von-Koeller** [comment]
@wkpark

Do I understand correctly that this is ready for review?

**Comment by wkpark** [comment]
> @wkpark
> 
> Do I understand correctly that this is ready for review?

Yes. I think so.

minimal fixes have been cherry-picked and squashed. (several unrelated hunks were carefully removed/simplified manually)

**Review by younesbelkada — APPROVED** [review]
Thanks @wkpark !
I think we should merge this PR before #908 right? Could you re-enable windows build in the workflow for #908 after we merge this PR?

**Comment by wkpark** [comment]
> Thanks @wkpark ! I think we should merge this PR before #908 right? Could you re-enable windows build in the workflow for #908 after we merge this PR?

right and thank you.
now I can add windows build matrix!

**Comment by younesbelkada** [comment]
Perfect thanks @wkpark !

**Comment by younesbelkada** [comment]
Thanks for your great work @wkpark I we can confirm with @Titus-von-Koeller that this PR did not break anything with respect to bnb + HF (PEFT, transformers, etc.) Looking forward to your next contributions!

**Pull request merged**